packages/inter-protocol/src/auction/auctioneer.js

```diff
@@ -0,0 +1,282 @@
+
+const { Fail, quote: q } = assert;
+
+const trace = makeTracer('Auction', false);
+
+const makeBPRatio = (rate, currencyBrand, collateralBrand = currencyBrand) =>
+  makeRatioFromAmounts(
+    AmountMath.make(currencyBrand, rate),
+    AmountMath.make(collateralBrand, 10000n),
+  );
+
+/**
+ * @param {ZCF<GovernanceTerms<{
+ *   StartFrequency: 'relativeTime',
+ *   ClockStep: 'relativeTime',
+ *   StartingRate: 'nat',
+ *   lowestRate: 'nat',
+ *   DiscountStep: 'nat',
+ * }> & {timerService: import('@agoric/time/src/types').TimerService, priceAuthority: PriceAuthority}>} zcf
+ * @param {{initialPoserInvitation: Invitation, storageNode: StorageNode, marshaller: Marshaller}} privateArgs
+ * @param {Baggage} baggage
+ */
+export const start = async (zcf, privateArgs, baggage) => {
+  const { brands, timerService: timer, priceAuthority } = zcf.getTerms();
+  timer || Fail`Timer must be in Auctioneer terms`;
+  const timerBrand = await E(timer).getTimerBrand();
+
+  const books = provideDurableMapStore(baggage, 'auctionBooks');
+  const deposits = provideDurableMapStore(baggage, 'deposits');
+  const brandToKeyword = provide(baggage, 'brandToKeyword', () =>
+    makeScalarBigMapStore('deposits', {
+      durable: true,
+    }),
+  );
+
+  const reserveFunds = provideEmptySeat(zcf, baggage, 'collateral');
+
+  const addDeposit = (seat, amount) => {
+    const depositListForBrand = deposits.get(amount.brand);
+    deposits.set(
+      amount.brand,
+      harden([...depositListForBrand, { seat, amount }]),
+    );
+  };
+
+  // Called "discount" rate even though it can be above or below 100%.
+  /** @type {NatValue} */
+  let currentDiscountRateBP;
+
+  const distributeProceeds = () => {
+    for (const brand of deposits.keys()) {
+      const book = books.get(brand);
+      const { collateralSeat, currencySeat } = book.getSeats();
+
+      const depositsForBrand = deposits.get(brand);
+      if (depositsForBrand.length === 1) {
+        // send it all to the one
+        const liqSeat = depositsForBrand[0].seat;
+
+        atomicRearrange(
+          zcf,
+          harden([
+            [collateralSeat, liqSeat, collateralSeat.getCurrentAllocation()],
+            [currencySeat, liqSeat, currencySeat.getCurrentAllocation()],
+          ]),
+        );
+        liqSeat.exit();
+        deposits.set(brand, []);
+      } else if (depositsForBrand.length > 1) {
+        const totCollDeposited = depositsForBrand.reduce((prev, { amount }) => {
+          return AmountMath.add(prev, amount);
+        }, AmountMath.makeEmpty(brand));
+
+        const collatRaise = collateralSeat.getCurrentAllocation().Collateral;
+        const currencyRaise = currencySeat.getCurrentAllocation().Currency;
+
+        const collShare = makeRatioFromAmounts(collatRaise, totCollDeposited);
+        const currShare = makeRatioFromAmounts(currencyRaise, totCollDeposited);
+        /** @type {import('@agoric/zoe/src/contractSupport/atomicTransfer.js').TransferPart[]} */
```

Comment:
Ugh, should not need a deep import just to use normal helper APIs. This is on me. Any idea how to export the type so it can be imported simply?

In any case, very cool use of a temporarily mutable array to gather all the parts of a rearrange!